### PR TITLE
Start a design spec for async commit

### PR DESCRIPTION
**Pull request by nrc**
This PR brings some of the project documentation up to date for async commit and starts a detailed technical description of the feature. There is still much work to do in the design spec, but I think we can land it now to make collaboration easier.

PTAL @sticnarf

### CODE REVIEW

**Comment by nrc** [comment]
I've added some more content about TiKV. PTAL @sticnarf @youjiali1995

File: design/async-commit/spec.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    uint64 min_commit_ts = 9;
+    repeated bytes secondaries = 10;
+}
+```
+
+## Clients
+
+Implemented in TiDB, TiSpark.
+
+The client has configuration for supporting async commit (`tikv-client.async-commit.enable`), supporting 1pc (`enable-one-pc`), forcing external consistency when using async commit (`external-consistency`), and limits on the size (`tikv-client.async-commit.total-key-size-limit`) and number of keys (`tikv-client.async-commit.keys-limit`) in an async commit transaction. When async commit is enabled, a transaction will not use async commit if it exceeds either of the size limits.
```
Comment (sticnarf) [inline design/async-commit/spec.md:71]: `async-commit.enable`, ``enable-one-pc` and `external-consistency` have been changed to system variables.

Comment (nrc) [reply design/async-commit/spec.md:71]: Sigh, spec out of date before it even lands :-(

**Comment by nrc** [comment]
@sticnarf @youjiali1995 updated, PTAL

**Review by youjiali1995 — APPROVED** [review]

**Review by sticnarf — COMMENTED** [review]
I think we can delay fallback related design until fallback is landed. The rest lgtm.

File: design/async-commit/spec.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+The server maintains a `max_ts`, which is the largest timestamp observed by the server. Note: the concurrency manager was implemented as part of the async commit project, however, here we only describe parts of it actually used for async commit.
+
+If `max_ts` has become out of date (due to some change to the region topology), then the store cannot handle the primary key of transactions using async commit, and must use full 2pc.
+
+### Prewrite
+
+The server iterates over all keys in the request. If a key is locked for async commit by the current transaction, and it is no longer an async commit transaction, then the lock can be overwritten. Each key is locked; for 1pc the lock is kept privately in memory, for 2pc the lock is written to storage. In either case the lock is also kept in memory by the concurrency manager. These in-memory locks are used to block reading while the transaction is live.
+
+The `min_commit_ts` for the key is calculated as the maximum of the transaction's start ts, `for_update_ts`, and `min_commit_ts`, and the `max_ts` of the store. The `min_commit_ts` is written to the lock. The largest `min_commit_ts` of all keys is returned to the client.
```
[the remark this answers is not in the document]
Comment (nrc) [reply design/async-commit/spec.md:95]: Hmm, I would expect to see some `+ 1`s here: https://github.com/tikv/tikv/blob/master/src/storage/txn/actions/prewrite.rs#L388-L405. Is it a mistake or are they somewhere else?

File: design/async-commit/spec.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+A lock on disk or in memory has the following new fields: `use_async_commit`, whether the lock belongs to an async commit transaction, `secondaries`, a list of secondary locks only used for the primary key, and `rollback_ts` (see below).
+
+The server maintains a `max_ts`, which is the largest timestamp observed by the server. Note: the concurrency manager was implemented as part of the async commit project, however, here we only describe parts of it actually used for async commit.
+
+If `max_ts` has become out of date (due to some change to the region topology), then the store cannot handle the primary key of transactions using async commit, and must use full 2pc.
+
+### Prewrite
+
+The server iterates over all keys in the request. If a key is locked for async commit by the current transaction, and it is no longer an async commit transaction, then the lock can be overwritten. Each key is locked; for 1pc the lock is kept privately in memory, for 2pc the lock is written to storage. In either case the lock is also kept in memory by the concurrency manager. These in-memory locks are used to block reading while the transaction is live.
```
[the remark this answers is not in the document]
Comment (nrc) [reply design/async-commit/spec.md:93]: I prefer to keep the spec a mirror of the code and we can update it when the new design lands.

[the remark this answers is not in the document]
Comment (sticnarf) [reply design/async-commit/spec.md:93]: OK.

File: design/async-commit/spec.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sticnarf) [reply design/async-commit/spec.md:95]: `.next()` is `+1`

[the remark this answers is not in the document]
Comment (nrc) [reply design/async-commit/spec.md:95]: ah, right, of course

**Pull request merged**